### PR TITLE
LTR MI cmdlets and tests 

### PR DESCRIPTION
**Pull request by xaliciayang**
## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by wyunchi-ms** [comment]
Hi @xaliciayang could you please take a look at the CI? The build is failed.

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

File: src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1 [hunk 1]
```diff
@@ -173,4 +173,151 @@ function Test-ManagedDeletedDatabaseShortTermRetentionPolicy
...
+#>
+function Test-ManagedInstanceLongTermRetentionPolicy($location = "southeastasia")
+{
+	# Setup
+	$resourceGroup = Create-ResourceGroupForTest
+	$resourceGroupName = $resourceGroup.ResourceGroupName
+	$managedInstance = Create-ManagedInstanceForTest
+	$managedInstanceName = $managedInstance.ManagedInstanceName
+	$weeklyRetention = "P1W"
+	$emptyRetention = "PT0S"
```
Comment (pixia) [inline src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1]: $zeroRetention

File: src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1 [hunk 2]
```diff
@@ -173,4 +173,151 @@ function Test-ManagedDeletedDatabaseShortTermRetentionPolicy
...
+	$backups = Get-AzSqlInstanceDatabase -ResourceGroup $resourceGroup -ManagedInstanceName $managedInstanceName -DatabaseName $databaseName | Get-AzSqlInstanceDatabaseLongTermRetentionBackup -OnlyLatestPerDatabase
+	Assert-AreNotEqual $backups.Count 0
+
+	# Restore Test
+	$backups = Get-AzSqlInstanceDatabaseLongTermRetentionBackup -Location $locationName
+	$db = Restore-AzSqlInstanceDatabase -FromLongTermRetentionBackup -ResourceId $backups[0].ResourceId -ResourceGroupName $resourceGroup -ManagedInstanceName $managedInstanceName -TargetDatabaseName $restoredDatabase
+	Assert-AreEqual $db.DatabaseName $restoredDatabase
+
+	# Test Remove with Piping
+	#Get-AzSqlDatabaseLongTermRetentionBackup -Location $locationName -ManagedInstanceName $serverName -DatabaseName $databaseWithRemovableBackup -BackupName $backups[0].BackupName | Remove-AzSqlDatabaseLongTermRetentionBackup
```
Comment (pixia) [inline src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1]: why is this commented out?

File: src/Sql/Sql.Test/Sql.Test.csproj [hunk 3]
```diff
@@ -26,4 +26,10 @@
     <ProjectReference Include="..\Sql\Sql.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="Microsoft.Azure.Management.Sql">
+      <HintPath>..\..\..\..\azure-sdk-for-net\artifacts\bin\Microsoft.Azure.Management.Sql\Debug\net452\Microsoft.Azure.Management.Sql.dll</HintPath>
```
Comment (pixia) [inline src/Sql/Sql.Test/Sql.Test.csproj]: need version change after .net

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/GetAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        /// </summary>
+        private const string GetBackupsByResourceIdSet = "GetBackupsByResourceId";
+
+        /// <summary>
+        /// The location the backups are in.
+        /// </summary>
+        [Parameter(Mandatory = true,
+            ParameterSetName = LocationSet,
+            Position = 0,
+            HelpMessage = "The location of the backups' source server.")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/GetAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: instance or managed instance, all places

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/GetAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        [Parameter(Mandatory = true,
+            ParameterSetName = GetBackupByResourceIdSet,
+            Position = 1,
+            HelpMessage = "The location of the backups' source server.")]
+        [Parameter(Mandatory = true,
+            ParameterSetName = GetBackupsByResourceIdSet,
+            Position = 1,
+            HelpMessage = "The location of the backups' source server.")]
+        [LocationCompleter("Microsoft.Sql/locations/longTermRetentionServers/longTermRetentionDatabases/longTermRetentonBackups")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/GetAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: update

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        /// </summary>
+        private const string RemoveBackupByResourceIdSet = "RemoveBackupByResourceId";
+
+        /// <summary>
+        /// Gets or sets the name of the location the backup is in.
+        /// </summary>
+        [Parameter(Mandatory = true,
+            ParameterSetName = RemoveBackupDefaultSet,
+            Position = 0,
+            HelpMessage = "The location of the backups' source server.")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: instance or managed instance

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        [ValidateNotNullOrEmpty]
+        [LocationCompleter("Microsoft.Sql/locations/longTermRetentionServers")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: Fix. longTermRetentionManagedInstances

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        public string ManagedInstanceName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the name of the database.
+        /// </summary>
+        [Parameter(Mandatory = true,
+            ParameterSetName = RemoveBackupDefaultSet,
+            Position = 2,
+            HelpMessage = "The name of the Managed Database the backup is from.")]
+        [ResourceNameCompleter("Microsoft.Sql/servers/databases", "ResourceGroupName", "ServerName")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: Fix Microsoft.Sql/servers/databases

File: src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            }
+            else if (!string.IsNullOrWhiteSpace(ResourceId))
+            {
+                ParseLongTermRentionBackupResourceId(ResourceId);
+            }
+
+            if (ShouldProcess(this.BackupName))
+            {
+                if (Force.IsPresent || ShouldContinue(
+                    string.Format(CultureInfo.InvariantCulture, Properties.Resources.RemoveAzureSqlDatabaseLongTermRetentionBackupDescription, this.BackupName, this.DatabaseName, this.ManagedInstanceName, this.Location),
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Cmdlet/RemoveAzureSqlManagedDatabaseLongTermRetentionBackup.cs]: I think you need a new constant (MI version) for RemoveAzureSqlDatabaseLongTermRetentionBackupDescription
RemoveAzureSqlDatabaseLongTermRetentionBackupWarning,

Please check all other places

File: src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseBackupLongTermRetentionPolicyModel.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+        ///// <summary>
+        ///// Construct AzureSqlManagedDatabaseBackupShortTermRetentionPolicyModel object
+        ///// </summary>
+        ///// <param name="resourceGroup">Resource group</param>
+        ///// <param name="managedInstanceName">Managed Instance name</param>
+        ///// <param name="managedDatabaseName">Managed Instance name</param>
+        ///// <param name="managedBackupRetentionPolicy">Managed Database object</param>
+        ///// <param name="deletionDate">Deletion date of the database, if it is deleted</param>
+        //public AzureSqlManagedDatabaseBackupLongTermRetentionPolicyModel(string resourceGroup, string managedInstanceName, string managedDatabaseName, string weeklyRetention, string monthlyRetention, string yearlyRetention, int? weekOfYear)
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseBackupLongTermRetentionPolicyModel.cs]: remove please, not used.

File: src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Gets or sets the resource ID.
+        /// </summary>
+        public string ResourceId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the server name.
+        /// </summary>
+        public string ServerName { get; set; }
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs]: Is this ServerName or InstanceName?

File: src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+        /// <summary>
+        /// Gets or sets the server create time.
+        /// </summary>
+        public DateTime? ServerCreateTime { get; set; }
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs]: Same, is this InstanceCreateTime?

Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs]: update description

File: src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+        /// </summary>
+        public string Location { get; set; }
+
+        /// <summary>
+        /// Gets or sets the resource ID.
+        /// </summary>
+        public string ResourceId { get; set; }
+
+        /// <summary>
+        /// Gets or sets the server name.
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs]: update description

File: src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupAdapter.cs [hunk 14]
```diff
@@ -201,5 +202,149 @@ internal AzureSqlDeletedManagedDatabaseBackupModel GetDeletedDatabaseBackup(stri
...
+        }
+
+        /// <summary>
+        /// Create or update a backup LongTermRetention policy for a Managed Database
+        /// </summary>
+        /// <param name="resourceGroup">The name of the resource group</param>
+        /// <param name="managedInstanceName">The name of the Azure SQL Server</param>
+        /// <param name="databaseName">The name of the Azure SQL Database</param>
+        /// <returns>A backup LongTermRetention policy</returns>
+        internal AzureSqlManagedDatabaseBackupLongTermRetentionPolicyModel SetManagedDatabaseBackupLongTermRetentionPolicy(
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupAdapter.cs]: Move this methods up to behind get-policy

File: src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupCommunicator.cs [hunk 15]
```diff
@@ -144,5 +144,142 @@ public RestorableDroppedManagedDatabase GetDeletedDatabaseBackup(string resource
...
+        /// <summary>
+        /// Get a backup LongTermRetention policy for a Azure SQL Database
+        /// </summary>
+        /// <param name="resourceGroup">The name of the resource group</param>
+        /// <param name="serverName">The name of the Azure SQL Server</param>
+        /// <param name="databaseName">The name of the Azure SQL Database</param>
+        /// <returns>A backup LongTermRetention policy</returns>
+        public ManagedInstanceLongTermRetentionPolicy GetManagedDatabaseLongTermRetentionPolicy(
+            string resourceGroupName,
+            string serverName,
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupCommunicator.cs]: instanceName

File: src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupCommunicator.cs [hunk 16]
```diff
@@ -144,5 +144,142 @@ public RestorableDroppedManagedDatabase GetDeletedDatabaseBackup(string resource
...
+                resourceGroupName,
+                serverName,
+                databaseName);
+        }
+
+        /// <summary>
+        /// Sets a database's Long Term Retention policy.
+        /// </summary>
+        /// <param name="resourceGroup">The resource group name.</param>
+        /// <param name="serverName">The server name.</param>
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase Backup/Services/AzureSqlManagedDatabaseBackupCommunicator.cs]: instanceName, please also update description

File: src/Sql/Sql/ManagedDatabase/Cmdlet/RestoreAzureSqlManagedDatabase.cs [hunk 17]
```diff
@@ -277,6 +293,11 @@ public class RestoreAzureRmSqlManagedDatabase
         [Parameter(ParameterSetName = GeoRestoreFromGeoBackupSetNameFromResourceIdParameterSet,
             Mandatory = true,
             HelpMessage = "The resource id of instance database object to restore")]
+        [Parameter(ParameterSetName =LtrBackupRestoreFromInputParametersSet,
+            Mandatory = true,
+            Position = 0,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "The resource id of the long term retention backup object to restore.")]
```
Comment (pixia) [inline src/Sql/Sql/ManagedDatabase/Cmdlet/RestoreAzureSqlManagedDatabase.cs]: long term retention manged instance backup

File: src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs (hunk 11 above)
Comment (xaliciayang) [reply src/Sql/Sql/ManagedDatabase Backup/Model/AzureSqlManagedDatabaseLongTermRetentionBackupModel.cs]: should be ManagedInstanceName

**Comment by wyunchi-ms** [comment]
Hi @xaliciayang , do you need any help with the CI?

File: src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1 [hunk 18]
```diff
@@ -173,4 +173,150 @@ function Test-ManagedDeletedDatabaseShortTermRetentionPolicy
...
+	.SYNOPSIS
+	Test long term retention backup commands for managed databases.
+#>
+function Test-ManagedInstanceLongTermRetentionBackup
+{
+
+	# MANUAL INSTRUCTIONS
+	# Create a server and database and fill in the appropriate information below
+	# Set the weekly retention on the database so that the first backup gets picked up, for example:
+	# Set-AzSqlInstanceDatabaseBackupLongTermRetentionPolicy -ResourceGroup $resourceGroup -ServerName $serverName -DatabaseName $databaseName -WeeklyRetention P1W
```
Comment (pixia) [inline src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1]: update cmdlet in instruction to MI version

File: src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1 [hunk 19]
```diff
@@ -173,4 +173,150 @@ function Test-ManagedDeletedDatabaseShortTermRetentionPolicy
...
+	# Wait about 18 hours until it gets properly copied and you see the backup when run get backups, for example:
+	# Get-AzSqlInstanceDatabaseLongTermRetentionBackup -Location $locationName -InstanceName $managedInstanceName -DatabaeName $databaseName
```
Comment (pixia) [inline src/Sql/Sql.Test/ScenarioTests/ManagedDatabaseBackupTests.ps1]: update

File: src/Sql/Sql/Properties/Resources.resx [hunk 20]
```diff
@@ -589,4 +589,10 @@
   <data name="CannotChangeHardwareFamily" xml:space="preserve">
     <value>You cannot change hardware family.</value>
   </data>
+  <data name="RemoveAzureSqlInstanceDatabaseLongTermRetentionBackupDescription" xml:space="preserve">
+    <value>Permanantly removing the Long Term Retention backup '{0}' on database '{1}' on server '{2}' in location '{3}'?</value>
```
Comment (pixia) [inline src/Sql/Sql/Properties/Resources.resx]: on managed instance '{2}'

File: src/Sql/Sql/Properties/Resources.resx [hunk 21]
```diff
@@ -589,4 +589,10 @@
...
+  </data>
+  <data name="RemoveAzureSqlInstanceDatabaseLongTermRetentionBackupWarning" xml:space="preserve">
+    <value>Are you sure you want to remove the Long Term Retention backup '{0}' on database '{1}' on server '{2}' in location '{3}'?</value>
```
Comment (pixia) [inline src/Sql/Sql/Properties/Resources.resx]: on managed instance '{2}'

File: src/Sql/Sql/ChangeLog.md [hunk 22]
```diff
@@ -19,9 +19,15 @@
...
 * Added PublicNetworkAccess to New-AzSqlServer and Set-AzSqlServer
+* Added support for Long Term Retention services for Managed instances
+    - Get/Set LTR policy on a managed database 
+    - Get LTR backup(s) by managed database, managed instance, or by location 
+    - Remove an LTR backup 
+    - Restore an LTR backup to create a new managed database
 
 ## Version 2.3.0
 * Added support for cross subscription point in time restore on Managed Instances.
+* Add support for changing existing Sql Managed Instance hardware generation
```
Comment (pixia) [inline src/Sql/Sql/ChangeLog.md]: remove

File: src/Sql/Sql/ChangeLog.md [hunk 23]
```diff
@@ -19,9 +19,15 @@
 -->
 ## Upcoming Release
 * Added PublicNetworkAccess to New-AzSqlServer and Set-AzSqlServer
+* Added support for Long Term Retention services for Managed instances
```
Comment (pixia) [inline src/Sql/Sql/ChangeLog.md]: Added support to manage Long Term Retention backups for Managed instances

File: src/Sql/Sql/help/Restore-AzSqlInstanceDatabase.md [hunk 24]
```diff
@@ -94,6 +94,13 @@ Restore-AzSqlInstanceDatabase [-FromGeoBackup] [-ResourceGroupName] <String> [-I
...
 
+### LongTermRetentionBackupRestoreParameter
+```
+Restore-AzSqlInstanceDatabase [-FromLongTermRetentionBackup] [-SubscriptionId <String>] [-ResourceId] <String>
+ -TargetInstanceDatabaseName <String> -TargetInstanceName <String> -TargetResourceGroupName <String> [-AsJob]
+ [-DefaultProfile <IAzureContextContainer>] [-WhatIf] [-Confirm] [<CommonParameters>]
+```
+
 ## DESCRIPTION
 The **Restore-AzSqlInstanceDatabase** cmdlet restores an instance database from a geo-redundant backup or a point in time in a live database.
```
Comment (pixia) [inline src/Sql/Sql/help/Restore-AzSqlInstanceDatabase.md]: The **Restore-AzSqlInstanceDatabase** cmdlet restores an instance database from a geo-redundant backup or a long term retention backup or a point in time in a live database.

Comment (xaliciayang) [reply src/Sql/Sql/help/Restore-AzSqlInstanceDatabase.md]: oops

**Comment by wyunchi-ms** [comment]
Hi @xaliciayang 
1. please donot remove `tools/PS-VSPrompt.lnk`.

**Pull request merged**